File: 2020/python/07.py
import operator
import re
from collections import defaultdict
from functools import reduce

from aocd import data

rg = defaultdict(list)
g = defaultdict(list)

for line in data.split("\n"):
    a = re.match(r"^(.*) bags contain", line)[1]
    for cnt, b in re.findall(r"(\d+) ([a-z ]+) bags?", line):
        rg[b].append(a)
        g[a].append((b, int(cnt)))

p1 = lambda node: reduce(operator.or_, map(p1, rg[node]), {node})
print(len(p1('shiny gold')) - 1)

p2 = lambda node: 1 + sum(p2(u) * cnt for u, cnt in g[node])
print(p2('shiny gold') - 1)

File: 2020/ruby/07.rb
require './utils.rb'

rg = Hash.new{|h,k| h[k] = []}
g = Hash.new{|h,k| h[k] = []}

read(7).lines.each{|line|
  a = line.match(/^(.*) bags contain/)[1]
  line.scan(/(\d+) ([a-z ]+) bags?/).each{|cnt,b|
    rg[b] << a
    g[a] << [b, cnt.to_i]
  }
}

p1 = ->n{rg[n].map{|x|p1[x]}.inject([n],:|)}
p p1['shiny gold'].size - 1

p2 = ->n{1+g[n].map{|u,c|p2[u]*c}.sum}
p p2['shiny gold'] - 1
